### DISCUSSION TITLE
How to override shouldComponentUpdate or prevent re-render when component invisible

### DISCUSSION THREAD

**Discussion by code-brewer** [mobxjs/mobx#2665] [Q&A] [answered]
Hi, all
    
   What I want to archive is prevent render action when the component is invisible.

   So my first thought is to override the `shouldComponentUpdate` function, as shown in below snippet.

  But it failed to work:  when store data change, the shouldComponentUpdate not called, and render not invoked either.

```
@inject('myStore')
@observer
class MyDashboardScreen extends Component {
    constructor(pros){
        super(pros);
    }

    shouldComponentUpdate() {
        const oldRet = super.shouldComponentUpdate ? super.shouldComponentUpdate() : null

        // this.state.isVisible : this state was set by withNavigationFocus HOC provide by reactnavigation
        //  reactnavigation site:  https://reactnavigation.org/docs/3.x/with-navigation-focus/ 

        let fixRet = oldRet && (this.state.isVisible)
        return fixRet
    }
    
  ...
}
```

Any ideas?   

Thanks for your attention~

**Comment by urugator** [comment, marked as answer]
https://github.com/mobxjs/mobx-react/issues/667#issuecomment-483747386

**Reply by code-brewer** [reply]
Thanks for your solution. 

Is there example on how to use untracked() in a decorated ES6  style class component ? 

  

> @inject('myStore')
@observer
class MyDashboardScreen extends Component {
   render() {   //  what kind of syntax to use  untracked()  ???
   }
  ...
}

**Reply by mweststrate** [reply]
It's the same, just put the code in your `render`

**Reply by urugator** [reply]
It's the same, you just have to wrap the rendering logic inside `untracked`, eg:
```javascript
class MyDashboardScreen extends Component {
   _render() {
      // original render
   }
   render() {
       return this.state.isVisible 
          ? this._render() 
          : untracked(() => this._render())
   }
}
```
Keep in mind, it only disables subscriptions for observables accessed directly inside `untracked`, it won't affect descendant observer components.

**Reply by code-brewer** [reply]
Very clear example and useful reminds！
